teuthology/dispatcher/__init__.py

```diff
@@ -0,0 +1,159 @@
+    os.execv(sys.executable, args)
+
+
+def stop():
+    log.info('Stopping...')
+    sys.exit(0)
+
+
+def load_config(archive_dir=None):
+    teuth_config.load()
+    if archive_dir is not None:
+        if not os.path.isdir(archive_dir):
+            sys.exit("{prog}: archive directory must exist: {path}".format(
+                prog=os.path.basename(sys.argv[0]),
+                path=archive_dir,
+            ))
+        else:
+            teuth_config.archive_base = archive_dir
+
+
+def main(args):
+    # run dispatcher in job supervisor mode if --supervisor passed
+    if args["--supervisor"]:
+        return supervisor.main(args)
+
+    verbose = args["--verbose"]
+    tube = args["--tube"]
+    log_dir = args["--log-dir"]
+    archive_dir = args["--archive-dir"]
+
+    # setup logging for disoatcher in {log_dir}
+    loglevel = logging.INFO
+    if verbose:
+        loglevel = logging.DEBUG
+    log.setLevel(loglevel)
+    log_file_path = os.path.join(log_dir, 'dispatcher.{tube}.{pid}'.format(
+        pid=os.getpid(), tube=tube))
+    setup_log_file(log_file_path)
+    install_except_hook()
+
+    load_config(archive_dir=archive_dir)
+
+    connection = beanstalk.connect()
+    beanstalk.watch_tube(connection, tube)
+    result_proc = None
+
+    if teuth_config.teuthology_path is None:
+        fetch_teuthology('master')
+    fetch_qa_suite('master')
+
+    keep_running = True
+    while keep_running:
+        # Check to see if we have a teuthology-results process hanging around
+        # and if so, read its return code so that it can exit.
+        if result_proc is not None and result_proc.poll() is not None:
+            log.debug("teuthology-results exited with code: %s",
+                      result_proc.returncode)
+            result_proc = None
+
+        if sentinel(restart_file_path):
+            restart()
+        elif sentinel(stop_file_path):
+            stop()
+
+        load_config()
+
+        job = connection.reserve(timeout=60)
+        if job is None:
+            continue
+
+        # bury the job so it won't be re-run if it fails
+        job.bury()
+        job_id = job.jid
+        log.info('Reserved job %d', job_id)
+        log.info('Config is: %s', job.body)
+        job_config = yaml.safe_load(job.body)
+        job_config['job_id'] = str(job_id)
+
+        if job_config.get('stop_worker'):
+            keep_running = False
+
+        job_config, teuth_bin_path = prep_job(
+            job_config,
+            log_file_path,
+            archive_dir,
+        )
+
+        # lock machines but do not reimage them
+        if 'roles' in job_config:
+            job_config = lock_machines(job_config)
+
+        run_args = [
+            os.path.join(teuth_bin_path, 'teuthology-dispatcher'),
+            '--supervisor',
+            '-v',
+            '--bin-path', teuth_bin_path,
+            '--archive-dir', archive_dir,
+        ]
+
+        with tempfile.NamedTemporaryFile(prefix='teuthology-dispatcher.',
```

Comment:
Instead of temporary file can we create this file in the job archive_dir for debugging, reproducibility and simplifying issue analysis.

Reply:
We create the same file (`orig.config.yaml`) in the job archive directory when `teuthology` cmd is invoked by the supervisor while initiating run in run.py [here](https://github.com/ceph/teuthology/blob/master/teuthology/run.py#L42). I'll refactor so that the file is created here.

Reply:
Right, just store under different name, no need to refactor teuthology.run right now, it would be better to drop saving the `orig.config.yaml` in follow up PR after teuthology-dispatcher. Another reason to minimize invasion to teuthology.run is that some system can still use older py2 branch for running teuthology test, so we want to have teuthology-worker to be backwards compatible for some time.

Reply:
Created `orig.config.yaml` in dispatcher itself (along with creating archive dirs for run and job) without refactoring run.py to maintain backward compatibility.